File: phonedesk-pc/relay_api.py
from __future__ import annotations

import base64
import hashlib
import json
import secrets
import time
from dataclasses import dataclass

import httpx

from identity import DeviceIdentityStore


def canonical_json(data: dict) -> bytes:
    return json.dumps(data, sort_keys=True, separators=(",", ":"), ensure_ascii=False).encode("utf-8")


def canonical_request(method: str, path: str, timestamp: str, nonce: str, body: bytes) -> bytes:
    body_hash = hashlib.sha256(body).hexdigest()
    return f"{method.upper()}\n{path}\n{timestamp}\n{nonce}\n{body_hash}".encode("utf-8")


@dataclass(frozen=True)
class DeviceRegistration:
    device_id: str
    display_name: str
    platform: str
    fingerprint: str
    public_key_der_b64: str


@dataclass(frozen=True)
class TrustedDeviceSummary:
    device_id: str
    display_name: str
    platform: str
    fingerprint: str
    public_key_der_b64: str


class RelayApi:
    def __init__(
        self,
        base_url: str,
        identity_store: DeviceIdentityStore,
        display_name: str,
        transport: httpx.BaseTransport | None = None,
        timeout: float = 15.0,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.identity_store = identity_store
        self.display_name = display_name.strip() or "Office Laptop"
        self._client = httpx.Client(base_url=self.base_url, transport=transport, timeout=timeout)
        self._device_id: str | None = None

    @property
    def device_id(self) -> str | None:
        return self._device_id

    def register(self) -> DeviceRegistration:
        identity = self.identity_store.load_or_create()
        body = canonical_json(
            {
                "display_name": self.display_name,
                "platform": "windows",
                "public_key_der_b64": base64.b64encode(identity.public_key_der).decode("ascii"),
            }
        )
        response = self._client.post(
            "/v1/devices/register",
            content=body,
            headers={"Content-Type": "application/json"},
        )
        response.raise_for_status()
        payload = response.json()
        registration = DeviceRegistration(
            device_id=payload["device_id"],
            display_name=payload.get("display_name", self.display_name),
            platform=payload.get("platform", "windows"),
            fingerprint=payload.get("fingerprint", identity.fingerprint),
            public_key_der_b64=payload.get(
                "public_key_der_b64",
                base64.b64encode(identity.public_key_der).decode("ascii"),
            ),
        )
        self._device_id = registration.device_id
        return registration

    def _ensure_registered(self) -> str:
        if self._device_id is None:
            self.register()
        assert self._device_id is not None
        return self._device_id

    def _signed_headers(self, method: str, path: str, body: bytes) -> dict[str, str]:
        device_id = self._ensure_registered()
        timestamp = str(int(time.time()))
        nonce = secrets.token_hex(16)
        message = canonical_request(method, path, timestamp, nonce, body)
        signature = self.identity_store.sign(message)
        return {
            "X-PhoneDesk-Device": device_id,
            "X-PhoneDesk-Timestamp": timestamp,
            "X-PhoneDesk-Nonce": nonce,
            "X-PhoneDesk-Signature": base64.b64encode(signature).decode("ascii"),
        }

    def _post_signed(self, path: str, payload: dict) -> httpx.Response:
        body = canonical_json(payload)
        headers = self._signed_headers("POST", path, body)
        headers["Content-Type"] = "application/json"
        response = self._client.post(path, content=body, headers=headers)
        response.raise_for_status()
        return response

    def _get_signed(self, path: str) -> httpx.Response:
        body = b""
        response = self._client.get(path, headers=self._signed_headers("GET", path, body))
        response.raise_for_status()
        return response

    def claim_pairing(self, pairing_id: str, manual_code: str) -> str:
        path = f"/v1/pairings/{pairing_id}/claim"
        response = self._post_signed(path, {"manual_code": manual_code})
        return str(response.json()["state"])

    def list_trusted(self) -> list[TrustedDeviceSummary]:
        response = self._get_signed("/v1/trusted-devices")
        devices = response.json().get("devices", [])
        return [
            TrustedDeviceSummary(
                device_id=item["device_id"],
                display_name=item["display_name"],
                platform=item.get("platform", "android"),
                fingerprint=item["fingerprint"],
                public_key_der_b64=item["public_key_der_b64"],
            )
            for item in devices
        ]

    def close(self) -> None:
        self._client.close()

File: phonedesk-android/app/src/main/java/com/phonedesk/remote/RelayApi.kt
package com.phonedesk.remote

import android.os.Build
import com.phonedesk.identity.DeviceIdentityStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.Base64
import java.util.UUID

class RelayApi(
    private val baseUrl: String,
    private val identityStore: DeviceIdentityStore,
    private val displayName: String = Build.MODEL.ifBlank { "Android phone" },
) : PairingRelayApi {
    private var deviceId: String? = null

    override suspend fun registerDevice(): DeviceRegistration {
        val identity = identityStore.getOrCreate()
        val body = RelayAuth.canonicalJson(
            mapOf(
                "display_name" to displayName,
                "platform" to "android",
                "public_key_der_b64" to Base64.getEncoder().encodeToString(identity.publicKeyDer),
            ),
        )
        val response = execute("POST", "/v1/devices/register", body, signed = false)
        val json = JSONObject(response)
        return DeviceRegistration(
            deviceId = json.getString("device_id"),
            fingerprint = json.getString("fingerprint"),
        ).also { registration -> deviceId = registration.deviceId }
    }

    override suspend fun createPairing(): PairingOffer {
        ensureRegistered()
        val response = execute("POST", "/v1/pairings", RelayAuth.canonicalJson(emptyMap()), signed = true)
        val json = JSONObject(response)
        return PairingOffer(
            pairingId = json.getString("pairing_id"),
            manualCode = json.getString("manual_code"),
            expiresInSeconds = json.getInt("expires_in_seconds"),
        )
    }

    override suspend fun getPairingStatus(pairingId: String): PairingCandidate {
        ensureRegistered()
        val path = "/v1/pairings/$pairingId"
        val json = JSONObject(execute("GET", path, ByteArray(0), signed = true))
        if (json.isNull("computer")) {
            throw IllegalStateException("No computer has claimed this pairing yet")
        }
        val computer = json.getJSONObject("computer")
        return PairingCandidate(
            deviceId = computer.getString("device_id"),
            displayName = computer.getString("display_name"),
            fingerprint = computer.getString("fingerprint"),
            publicKeyDerBase64 = computer.getString("public_key_der_b64"),
        )
    }

    override suspend fun confirmPairing(pairingId: String, approved: Boolean) {
        ensureRegistered()
        execute(
            "POST",
            "/v1/pairings/$pairingId/confirm",
            RelayAuth.canonicalJson(mapOf("approved" to approved)),
            signed = true,
        )
    }

    private suspend fun ensureRegistered(): String {
        if (deviceId == null) registerDevice()
        return requireNotNull(deviceId)
    }

    private suspend fun execute(
        method: String,
        path: String,
        body: ByteArray,
        signed: Boolean,
    ): String = withContext(Dispatchers.IO) {
        val connection = URL(baseUrl.trimEnd('/') + path).openConnection() as HttpURLConnection
        try {
            connection.requestMethod = method
            connection.connectTimeout = 15_000
            connection.readTimeout = 15_000
            connection.setRequestProperty("Accept", "application/json")

            if (signed) {
                val currentDeviceId = requireNotNull(deviceId) { "PhoneDesk device is not registered" }
                val timestamp = (System.currentTimeMillis() / 1000L).toString()
                val nonce = UUID.randomUUID().toString().replace("-", "")
                val canonical = RelayAuth.canonicalRequest(method, path, timestamp, nonce, body)
                val signature = identityStore.sign(canonical)
                connection.setRequestProperty("X-PhoneDesk-Device", currentDeviceId)
                connection.setRequestProperty("X-PhoneDesk-Timestamp", timestamp)
                connection.setRequestProperty("X-PhoneDesk-Nonce", nonce)
                connection.setRequestProperty(
                    "X-PhoneDesk-Signature",
                    Base64.getEncoder().encodeToString(signature),
                )
            }

            if (method == "POST") {
                connection.doOutput = true
                connection.setRequestProperty("Content-Type", "application/json")
                connection.outputStream.use { stream -> stream.write(body) }
            }

            val status = connection.responseCode
            val stream = if (status in 200..299) connection.inputStream else connection.errorStream
            val response = stream?.bufferedReader(Charsets.UTF_8)?.use { it.readText() }.orEmpty()
            if (status !in 200..299) {
                throw IllegalStateException("Relay request failed ($status): ${response.take(180)}")
            }
            response
        } finally {
            connection.disconnect()
        }
    }
}
